### ISSUE TITLE
Questions about development environment

### ISSUE THREAD

**Issue by mrclary** [spyder-ide/spyder#17406]
@spyder-ide/core-developers, I have some questions about the recommended development environment (see [`CONTRIBUTING.md`](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)).

Spyder's recommended development environment (`spyder-dev`) consists of packages installed from `requirements/conda.txt`, which include release versions of `python-lsp-server`, `spyder-kernels`, and `qdarkstyle`. Then `bootstrap.py` installs `python-lsp-server` from `external-deps` directory in develop mode and the executable is installed in a dedicated directory. Rather than installing `spyder-kernels` and `qdarkstyle` in develop mode, `bootstrap.py` adds them to `sys.path`. Additionally, rather than install `spyder` in develop mode, `bootstrap.py` manually adds plugin entry points and a fake `spyder` distribution to `pkg_resources` and adds the repo to `sys.path`.

So, my questions are
1. Why install `python-lsp-server` in develop mode but not `spyder-kernels` or `qdarkstyle`?
2. Why install `pylsp` executable in a dedicated directory instead of the standard installation location?
3. Why not install the local spyder repo in develop mode?

Since `spyder-dev` is a dedicated development environment, wouldn't it be better practice to install `pylsp` in a standard location, and install `spyder-kernels` and `qdarkstyle` in develop mode as well? Then `sys.path` does not have to be tampered with, which may be less robust than installing in develop mode. Additionally, note that `bootstrap.py` redundantly adds `python-lsp-server` to `sys.path` since setuptools adds it during develop mode installation. Also, wouldn't installing `spyder` in develop mode would be a more robust way to ensure that all the entry points are made via the egg-info?

**Comment by CAM-Gerlach** [comment]
For a while, I've been thinking that a lot (though not necessarily all) of what the `bootstrap.py  script does is not really necessary to have a custom script do in the modern era, and could be replaced by installing in development mode (either via the legacy Setuptools-specific method or the new standardized PEP 660 hooks, either invoked via `pip install -e ,`).

Aside from that script being written in an era before that was easily available or at least commonplace (though that doesn't explain some of the more recent additions you mention), one reason for having it not use `pip install -e .` is that it allows a separate copy of Spyder to be installed in the current environment, either from a built wheel or a production release, without overwriting it and switching back and forth. But I'm not sure how common it is nowadays that developers do that.

**Comment by mrclary** [comment]
> one reason for having it not use `pip install -e .` is that it allows a separate copy of Spyder to be installed in the current environment, either from a built wheel or a production release, without overwriting it and switching back and forth.

This reason also seems obsolete, as it seems to defeat the purpose of a dedicated environment. Conda and other virtual environments are so easily created/destroyed, and since the purpose of a dedicated environment is to ensure that there is no accidental cross-contamination (multiple versions of spyder installed in the same environment may be asking for trouble!), wouldn't be best practice to have separate environments for separate versions?

>  But I'm not sure how common it is nowadays that developers do that.

Exactly.

Additionally, `bootstrap.py` removes `python-lsp-server`'s egg-info file on exit. This is problematic if multiple environments have that package installed in develop mode, such as an environment for building the macOS app, which rely on the egg-info directory to be present.

**Comment by mrclary** [comment]
PR #17408 shows my proposed update to the bootstrap script.

**Comment by CAM-Gerlach** [comment]
> This reason also seems obsolete, 

Right; I used to do this years ago, but haven't for quite a while now, for the reasons you mention.

> Additionally, bootstrap.py removes python-lsp-server's egg-info file on exit. This is problematic if multiple environments have that package installed in develop mode, such as an environment for building the macOS app, which rely on the egg-info directory to be present.

FYI, `.egg-info` is a an implementation detail of the legacy Setuptools develop install method, and one of the last vestiges of the nearly obsolete and soon to finally be removed easy_install Egg binary package format. If relying on it specifically (as opposed to a develop install more generally) is indeed what you mean, then that's something we should probably prepare to rethink.

**Comment by mrclary** [comment]
> FYI, `.egg-info` is a an implementation detail of the legacy Setuptools develop install method, and one of the last vestiges of the nearly obsolete and soon to finally be removed easy_install Egg binary package format. If relying on it specifically (as opposed to a develop install more generally) is indeed what you mean, then that's something we should probably prepare to rethink.

I'm referring to a develop install more generally. However, at present, I believe, `py2app` relies on `pkg_resources` to gather the relevant packages to bundle in the macOS app. When `python-lsp-server`'s egg-info is removed, it and it's dependencies aren't found and bundled into the app.

**Comment by mrclary** [comment]
I suppose I should clarify. When installing in develop mode, a link to the local egg-info is placed in the standard site-packages location. So when the egg-info directory is removed, any environment that has an egg link to it, will not resolve it and break the environment. So I don't think it is the 'egg-info' file specifically, but the artifacts of the develop install are messed with that disrupt other environments.

**Comment by mrclary** [comment]
I cannot say with any certainty what the internal workings of `py2app` are, nor what their future development plans are. Nevertheless, on our end, I see no need to "uninstall" (however that's defined) the local `python-lsp-server` develop installation. The way I see it, one has a development environment, one installs the package locally in develop mode, and that's it. Of course, if I'm missing something here, or someone has a compelling use case for uninstalling it every time they run `bootstrap.py`, I'd like to enlightened.

**Comment by CAM-Gerlach** [comment]
> Nevertheless, on our end, I see no need to "uninstall" (however that's defined) the local python-lsp-server develop installation. 

Sure—I'm just clarifying that if there _is_ such a need, it should be done via `pip uninstall`, not the current hacky, deprecated-implementation-detail-specific approach, since I didn't yet see either change in the PR.

**Comment by ccordoba12** [comment]
I'm sorry, but I'm strongly -1 on changing the `bootstrap.py` behavior, even if it implies more maintenance burden (which I'm willing to do). The problems, as I see them, are that:

1. Tampering with the environment to install a bunch of stuff in development mode doesn't work well with conda, which can lead to a broken env quite easily (e.g. conda can uninstall all spyder-kernels dependencies because that package is not in the env).
2. People use `bootstrap.py` to simply run and test the latest Spyder, not to develop it. In that case, I see no need to leave their envs in very bad shape.
3. This would make contributing to Spyder more difficult because newbies don't understand what development mode is and works, and how it can affect their envs.

You can add an additional script to do what you want, if you prefer to develop in development mode, but I think the current way `bootstrap.py` works is just fine.

*Note*: I have very little time for these discussions right now, so I'll probably answer to your replies during the weekend or next week.

**Comment by mrclary** [comment]
@ccordoba12, thanks for your response.

> > Why install python-lsp-server in develop mode but not spyder-kernels or qdarkstyle?
> 
> Because `python-lsp-server` uses setuptools entry points for its plugins, so it needs to be installed somewhere for them to work properly. That's my understanding of the situation, but @andfoy was the one who implemented the solution and I think there's no other way around it.

Indeed, `python-lsp-server` does need to be installed; the question was why not also `spyder-kernels` and `qdarkstyle`.

> > Why install pylsp executable in a dedicated directory instead of the standard installation location?
> 
> To avoid touching people's environment.

But their environment is modified anyway. When installing `python-lsp-server` in develop mode, pip removes the existing installation and replaces it with an egg-link to the local repo, so a dedicated installation directory for the executable provides no benefit in this regard.

In response to your other comments
1. Adding `spyder`, `spyder-kernels`, and `qdarkstyle` local repo paths to `sys.path` does not provide any benefit over installing them in editable mode with `--no-deps`, as far as mixing pip and conda is concerned, because in both cases the local package will import from conda installed dependencies. However, installing in editable mode with `--no-deps` has the advantage of being more robust and avoids unnecessarily polluting `sys.path`.
2. If someone uses `bootstrap.py` to simply run and test the latest Spyder, the environment's `python-lsp-server` will be affected. At this point, it doesn't matter whether it was just `python-lsp-server` or additionally `spyder`, `spyder-kernels`, and `qdarkstyle`, because remedial action will be required to restore the environment.
3. I, too, would like to make it as easy as possible for newbies to contribute to Spyder. What circumstance do you envision a newbie to be in? The simplest scenario I can think of is they have Anaconda installed and are only using the base environment. They then read our contribution guide instructing them to clone the Spyder repo locally. The contributing guide also instructs them to create a `spyder-dev` environment. This is exactly what I would recommend to everyone of _any_ experience level, _especially_ if they are a newbie. Creating a separate environment is the _easiest_ way to prevent issues _particularly_ for inexperienced users.

**Comment by CAM-Gerlach** [comment]
> I'm sorry, but I'm strongly -1 on changing the bootstrap.py behavior, even if it implies more maintenance burden (which I'm willing to do).

For better or worse, you're not going to have much choice soon than to make at least a lot of these changes (or even more complex/hacky ones), since from a packaging perspective a number of the things the script and lsp utils is currently doing have been discouraged and deprecated for many, many years (thus it is already generating DeprecationWarnings) and are on track to be removed very soon. Furthermore, the hacks used here won't, in general, work with the modern PEP 517/pyproject.toml-based ecosystem, and could cause much more serious breakage to the environment as a whole than doing things the recommended way.

> 1. Tampering with the environment to install a bunch of stuff in development mode doesn't work well with conda, which can lead to a broken env quite easily (e.g. conda can uninstall all spyder-kernels dependencies because that package is not in the env).

TL;DR: What you're doing now is strictly _more_ dangerous, both with and without Conda, than what is proposed here, and most of it has been deprecated for years and will break completely very soon.

I don't really follow your reasoning here, as it applies much more strongly to the manual "tampering" with the env via hacks previously used to try to fake a package install than just doing a proper `pip install -e --no-deps`. While the interaction can get complicated if not handled carefully, Conda _does_ recognize and generally respect pip-installed packages, whereas if it isn't installed at all, of course it will not.

Even so, it doesn't uninstall things it is not told to except in specific user-commanded circumstances, and certainly not if the dependencies have actually been specified in the conda requirements.txt (i.e. they were user-specified package specs passed to `conda install`).

Furthermore, editable installs with pip (and occasionally even tightly coupled dependencies as well) is the standard way packages are developed with conda, and is what I've done in dozens of other environments for years now without any of these problems. Could you be more specific here about what particular issues you've faced, and with what environment setup specifically?

To the contrary, I've heard and seen firsthand all the things that can go wrong with manually trying to hack `sys.path`, use long-deprecated Egg functionality, rely on executing `setup.py` directly, legacy Setuptools `develop` install, etc. And again, all or almost all of those things are actively being removed very soon and will break completely in the very near future, so you'll have to do what we suggest anyway (or come up with an even more complicated and hackier alternative).

> People use bootstrap.py to simply run and test the latest Spyder, not to develop it. In that case, I see no need to leave their envs in very bad shape.

We are already telling them to create an environment specifically for this, which is the rule with conda, having them install our deps in that environment, and then hacking that environment to install PyLSP. Their environment is going to be in better shape (such that they can actually run Spyder by the standard mechanisms) with this script than doing aweful manual hacks that will break soon.

> This would make contributing to Spyder more difficult because newbies don't understand what development mode is and works, and how it can affect their envs.

*editable mode; it only affects the env we specifically told them to create for Spyder development, and otherwise at a basic level, has essentially the same user-facing behavior as the existing bootstrap script, just without all the aweful, likely to break hacks behind the scenes, and with more functionality if they choose to use it. Furthermore, this is the install method that essentially every single other package uses for development use, so it adds more confusion that we're using bespoke hacks that don't work as people expect.

@mrclary 

> Indeed, python-lsp-server does need to be installed; the question was why not also spyder-kernels and qdarkstyle.

Because, as @ccordoba12 mentioned, the LSP entry points needed to be available. However, regardless, I still agree that since we're modifying the user's env anyway with this, we may as well do it for the others, and do it properly and not use horrible hacks that are more likely to break things.

**Comment by ccordoba12** [comment]
I missed that this one was fixed by PR #17408, so I'm going to close it.
